[WIP] Abort in TimerOutput::~Scope on uncaught exception

gassmoeller opened the pull request:
This is an attempt at a better fix for geodynamics/aspect#4033. In short we trigger MPI comunication in the destructor of a TimerOutput::Scope (if the timer is created within an MPI context), but if the destructor is called because an exception is thrown that may mean other processes are doing different MPI calls at the same time, producing weird error messages or deadlocks. 

In order to work around this we could MPI_Abort if an exception is uncaught in the destructor, I took the code for that from #11444.

I can confirm that this fixes the problems we had in geodynamics/aspect#4033, i.e. if I run an unpatched ASPECT with this patched version of deal.II I run into the MPI_Abort  message correctly. However, for some reason I can not provoke the same behavior in the test I attached and need to do more testing on that. Currently the test simply throws the exception, the error message does not trigger (nor does the MPI_Abort) and I also do not get a deadlock if I remove the patch and run the test. Before I get this test working I would not want to move forward with this. Does someone have an idea why the exception in the test simply gets thrown like normal with no MPI deadlock? Ignore the test output for now.

@bangerth: Is this approximately what you were talking about at last week's meeting? Or should the check for an active exception move into `Timer::stop()`, which is the function that will do the call to `dealii::Utilities::MPI::min_max_avg`, which causes the issues?

bangerth commented on source/base/timer.cc:
```diff
@@ -1042,8 +1042,51 @@ TimerOutput::reset()
+    if (std::uncaught_exception() == true)
+#  endif
+      {
+        std::cerr
```
This shouldn't happen anyway. If we get here, there's a bug in the code. So I think that's ok.

bangerth commented on source/base/timer.cc:
```diff
@@ -1042,8 +1042,51 @@ TimerOutput::reset()
+   * message).
+   */
+  void
+  check_exception(const std::string &scope_name)
```
I know that the place where you copied this from calls this `check_exception`, but would you mind renaming this function `abort_on_pending_exception`?

bangerth reviewed (commented):
You will apparently have to re-run the indentation script. I thought I might as well tack on a naming request.

bangerth commented:
Separately, though: The test does not actually show any specific output for this functionality, right? Or is it the fact that it doesn't hang that is specific?

bangerth commented:
@gassmoeller Ping? If you want to get this into 9.3, it would have to happen tomorrow :-)

tjhei commented:
Can you two help me understand this proposed change? If I understand it correctly, I would no longer be allowed to let an exception propagate past a Scope. I in fact do this in one of my codes (not public):
Solve() contains a Scope but can throw SolverNoConvergence, which I catch in the main run function and handle it gracefully. As it is thrown on all ranks, this works as expected. Maybe not elegant, but it worked so far.
How would I fix this without big changes?

bangerth commented:
That's a valid point. I had thought about this when reviewing and hoped that nobody does that, but we do these sorts of things in ASPECT as well where we propagate exceptions in solvers at least some distance. In a sense, the fact that we have a `TimerOutput::Scope` object is an indication that this kind of design is indeed encouraged.

So at the very least, let's defer this PR until after the release so we have more time to think about this. 

Longer term, one could think about something like this: There are ways to provide white-lists or black-lists of exception types via the mechanisms of https://en.cppreference.com/w/cpp/error/current_exception that one could use to call the `MPI_Abort` if, for example, the exception is not one of a list of types. So something of the kind
```
  class TimerOutput {
    ...

    template <typename... T>             // variadic template, 
    void trigger_abort_in_destructor_if_not_one_of ();

    ...
  };

  TimerOutput::Scope::~Scope ()
  {
    if (uncaught exception)
      if (type of exception not in list of types previously set)
        MPI_Abort();
  }
```

bangerth commented:
On 5/21/21 10:37 AM, Rene Gassmoeller wrote:
> I unified the error handling between ~TimerOutput and ~Scope (I now use the 
> same code as ~TimerOutput) so at least functionality wise they are now 
> behaving in the same way. However, I can not finish the test, because I have 
> to throw an exception in order to trigger the right behavior in the test and I 
> do not know how to produce test output for this. Is there a way to indicate 
> that throwing an exception and crashing (instead of deadlocking) is the right 
> outcome for this test? If I catch the exception in the test and not rethrow 
> |~MPI_InitFinalize| will deadlock, because it calls MPI_Finalize.

You could output something *after* that point. The fact that it doesn't show 
up in the output indicates that `MPI_Abort` was called.

bangerth commented:
On 5/21/21 10:37 AM, Rene Gassmoeller wrote:
> But even if we figure out a way around this, how would you ever be able to 
> recover from an exception on a single process in an MPI program, if you do not 
> know what MPI calls other processes have made in the meantime, and you cannot 
> find out, because using MPI could cause a deadlock? Would the correct policy 
> be to never allow MPI communication when an uncaught exception exists?

Things just need to be synchronous: Either all processors exit the scope via 
an exception, or none do. For example, when you call SolverCG::solve and it 
fails, then it will fail with an exception on all processors. That seems 
reasonable behavior to me.

gassmoeller commented:
I think before I continue working on this PR we should decide if usage like the one @tjhei mentioned is acceptable or not. If we decide that propagating exceptions out of timer scopes and catching them is intended, then there is nothing we can change on the current behavior. I am ok with that, it is a useful thing to do. However, sometimes it can lead to these weird MPI errors or deadlocks and then we need to remember it is likely an exception that is not thrown by all ranks (in the user code). I was hesitant to open this PR anyway, because once geodynamics/aspect#4033 is merged my issue is fixed anyway, independent of this PR.

tamiko commented:
I am not particularly in favor of adding such an exception guard into the `TimerOutput` class. What about we add a separate lightweight `MPIExceptionGuard` class (that could also take an exception object as a template parameter to print  an instructive error message)?

I would prefer such a setup because
 - (a) ensuring that we don't diverge on mpi ranks due to exceptions with such a guard class is an independently useful concept; and
 - (b), as pointed out by @tjhei, there are valid use cases of letting an exception propagate past a `TimerOutput`  scope.

tjhei commented on include/deal.II/base/timer.h:
```diff
@@ -563,7 +581,7 @@ class TimerOutput
 
     /**
-     * Destructor calls stop()
+     * Destructor calls stop().
```
```suggestion
     * Destructor calls stop_lap().
```
?

tjhei commented on include/deal.II/base/timer.h:
```diff
@@ -884,7 +902,12 @@ class TimerOutput
+
+  mutable bool is_synchronized;
+
+  void
```
add a comment

tjhei commented on source/base/timer.cc:
```diff
@@ -277,17 +272,56 @@ Timer::wall_time() const
+void
+Timer::synchronize() const
+{
+  if (is_synchronized == false)
```
does this need to be guarded to only execute if we have MPI?

tjhei commented:
> However, the solution is incomplete in the sense that users can still trigger deadlocks through user code, for example if they ask for timer output in a destructor of their classes (this will trigger communication, leading to a deadlock).

I think this is an acceptable situation.

> Are you ok with this approach?

I think this is a much better solution. :+1:

gassmoeller commented:
Ok, I think this is finally ready for another look. @tjhei or @bangerth could you do another review? Functionally this is still the same as the last version we discussed, it is just a lot simpler to read because I extracted many changes into other PRs.

I addressed all your comments, but I wasnt sure about this one:

> As a general rule, if you access mutable variables, you need a mutex to do so.

I dont think this statement is true in this form, but I do not know that much about thread-parallel programming. As far as I understand I need a mutex whenever I access (read or write) a variable that may be changed by another thread at the same time. This is not directly related to mutable variables, as non mutable variables can still be changed by non-const member functions on another thread. But it is more likely for mutable variables to be changed, because they can also be changed by const functions. Thus I get your point, I introduced a mutex to the function `start()`, `stop()`, and `synchronize_and_update()`, which should at least prevent multiple threads from writing to the variables at the same time. There are still scenarios, where one thread would be calling `cpu_time()`, while another stops a lap that would not work, but if we want to prevent that, we would need a mutex in every member function that relies on the state of the class (and generally every member function of every class that has a state that can be modified after construction?). I could do that by moving the mutex from `synchronize_and_update()` into all member functions, but is that what we want?
